src/Microsoft.AspNetCore.Razor.Language/RazorDiagnosticFactory.cs

```diff
@@ -422,6 +422,56 @@ public static RazorDiagnostic CreateTagHelper_InlineMarkupBlocksNotSupportedInAt
             return diagnostic;
         }
 
+        internal static readonly RazorDiagnosticDescriptor Parsing_EmptyTagHelperBoundAttribute =
+            new RazorDiagnosticDescriptor(
+                $"{DiagnosticPrefix}2008",
+                () => LegacyResources.RewriterError_EmptyTagHelperBoundAttribute,
+                RazorDiagnosticSeverity.Error);
+        public static RazorDiagnostic CreateParsing_EmptyTagHelperBoundAttribute(SourceSpan location, string attributeName, string tagName, string propertyTypeName)
+        {
+            return RazorDiagnostic.Create(Parsing_EmptyTagHelperBoundAttribute, location, attributeName, tagName, propertyTypeName);
+        }
+
+        internal static readonly RazorDiagnosticDescriptor TagHelper_IndexerAttributeNameMustIncludeKey =
+            new RazorDiagnosticDescriptor(
+                $"{DiagnosticPrefix}2009",
+                () => LegacyResources.TagHelperBlockRewriter_IndexerAttributeNameMustIncludeKey,
+                RazorDiagnosticSeverity.Error);
+        public static RazorDiagnostic CreateTagHelper_IndexerAttributeNameMustIncludeKey(SourceSpan location, string attributeName, string tagName)
+        {
+            return RazorDiagnostic.Create(TagHelper_IndexerAttributeNameMustIncludeKey, location, attributeName, tagName);
+        }
+
+        internal static readonly RazorDiagnosticDescriptor TagHelper_AttributeListMustBeWellFormed =
+            new RazorDiagnosticDescriptor(
+                $"{DiagnosticPrefix}2010",
+                () => LegacyResources.TagHelperBlockRewriter_TagHelperAttributeListMustBeWellFormed,
+                RazorDiagnosticSeverity.Error);
+        public static RazorDiagnostic CreateTagHelper_AttributeListMustBeWellFormed(SourceSpan location)
+        {
+            return RazorDiagnostic.Create(TagHelper_AttributeListMustBeWellFormed, location);
+        }
+
+        internal static readonly RazorDiagnosticDescriptor TagHelper_CannotHaveCSharpInTagDeclaration =
+            new RazorDiagnosticDescriptor(
+                $"{DiagnosticPrefix}2011",
+                () => LegacyResources.TagHelpers_CannotHaveCSharpInTagDeclaration,
+                RazorDiagnosticSeverity.Error);
+        public static RazorDiagnostic CreateTagHelper_CannotHaveCSharpInTagDeclaration(SourceSpan location, string tagName)
+        {
+            return RazorDiagnostic.Create(TagHelper_CannotHaveCSharpInTagDeclaration, location, tagName);
+        }
+
+        internal static readonly RazorDiagnosticDescriptor TagHelper_AttributesMustHaveAName =
```

Comment:
Should these also be called `Parsing_Blah` instead of `TagHelper_Blah`? 
Also, these are currently grouped under semantic errors. But we also have a region for TagHelper errors. I'm not sure where we should be putting TagHelper parsing errors.

Reply:
Ya you're right, these should all be Syntactic errors. That' what I get for continuing where I last added errors 😉. As for naming, I can update them to Parsing_TagHelper*

Reply:
`Parsing_EmptyTagHelperBoundAttribute` might be a semantic error though. What do you think?

Reply:
I could see that.